### PR TITLE
SNOW-787480: fix json load encoding error

### PR DESCRIPTION
**Pull request by sfc-gh-aling**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-787480

next action item: add telemetry to collect encoding info: https://snowflakecomputing.atlassian.net/browse/SNOW-801885

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am modifying authorization mechanisms
   - [ ] I am adding new credentials
   - [ ] I am modifying OCSP code
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/connector/version.py [hunk 1]
```diff
@@ -1,3 +1,3 @@
 # Update this for the versions
 # Don't change the forth version number from None
-VERSION = (3, 0, 3, None)
+VERSION = (3, 0, 4, None)
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/version.py:3]: Let's not change this until the next release?

File: DESCRIPTION.md [hunk 2]
```diff
@@ -8,6 +8,10 @@ Source code is also available at: https://github.com/snowflakedb/snowflake-conne
 
 # Release Notes
 
+- v3.0.4(Unreleased)
+
+  - Fixed a bug that JSON formatted returning data is decoded in wrong encoding.
```
Comment (sfc-gh-stan) [inline DESCRIPTION.md:13]:
```suggestion
  - Fixed a bug where query results data in JSON format is decoded with the wrong encoding.
```

Comment (sfc-gh-stan) [reply DESCRIPTION.md:13]: Shall we link https://github.com/chardet/chardet/issues/148?

File: test/integ/test_cursor.py [hunk 3]
```diff
@@ -1618,3 +1618,27 @@ def test_multi_statement_failure(conn_cnx):
...
+    # SNOW-787480, if not explicitly setting utf-8 encoding, the data will be
+    # detected encoding as windows-1250 by chardet.detect
+    # which is wrong, with the utf-8 fix, we can get the correct decoded data
+
+    local_result_batch = JSONResultBatch(None, None, None, None, None, None)
+    from snowflake.connector.vendored.requests import Response
+
+    resp = Response()
+    resp._content = '{"key": "á"}'.encode("latin1")
+    assert local_result_batch._load(resp) == [
```
[the remark this answers is not in the document]
Comment (sfc-gh-mkeller) [reply test/integ/test_cursor.py:1634]: I agree

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]

File: src/snowflake/connector/version.py (hunk 1 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/version.py]: reverted

File: src/snowflake/connector/connection.py [hunk 4]
```diff
@@ -204,6 +204,10 @@ def DefaultConverterClass() -> type:
         True,
         bool,
     ),  # Whether to log imported packages in telemetry
+    "json_result_force_utf8_decoding": (
+        False,
+        bool,
+    ),  # Whether to force the JSON content to be decoded in utf-8, it only works when result format is JSON
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/connection.py:210]: Let's not say "only works", what about "is only effective"?

File: src/snowflake/connector/connection.py [hunk 5]
```diff
@@ -265,6 +269,9 @@ class SnowflakeConnection:
         enable_connection_diag: when true, clients will generate a connectivity diagnostic report.
         connection_diag_log_path: path to location to create diag report with enable_connection_diag.
         connection_diag_whitelist_path: path to a whitelist.json file to test with enable_connection_diag.
+        json_result_force_utf8_decoding: When true, json result will be decoded in utf-8,
+          when false, the encoding of the content is auto-detected. Default value is false.
+          This parameter only works when the result format is JSON.
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/connection.py:274]: Ditto.

File: src/snowflake/connector/connection.py [hunk 6]
```diff
@@ -204,6 +204,10 @@ def DefaultConverterClass() -> type:
         True,
         bool,
     ),  # Whether to log imported packages in telemetry
+    "json_result_force_utf8_decoding": (
+        False,
```
Comment (sfc-gh-stan) [inline src/snowflake/connector/connection.py:208]: Just curious, why not enable by default?

File: test/integ/test_cursor.py [hunk 7]
```diff
@@ -1618,3 +1618,30 @@ def test_multi_statement_failure(conn_cnx):
...
             (type(None), str),
         )
+
+
+@pytest.mark.skipolddriver
+def test_decoding_utf8_for_json_result(conn_cnx):
+    # SNOW-787480, if not explicitly setting utf-8 decoding, the data will be
+    # detected decoding as windows-1250 by chardet.detect
+    with conn_cnx() as con, con.cursor() as cur:
+        cur.execute("alter session set python_connector_query_result_format='JSON'")
```
Comment (sfc-gh-mkeller) [inline test/integ/test_cursor.py:1628]:
```suggestion
    with conn_cnx(
        session_parameters={"python_connector_query_result_format": "json}
    ) as con, con.cursor() as cur:
```

File: test/integ/test_cursor.py [hunk 8]
```diff
@@ -1618,3 +1618,30 @@ def test_multi_statement_failure(conn_cnx):
...
+        # the backend behavior is flaky
+        assert ret[0] in (
+            ('"","","","","",OfigrĂ\xa0fic"",',),
+            ('"","","","","",Ofigràfic"",',),
+        )
+
+    with conn_cnx(json_result_force_utf8_decoding=True) as con, con.cursor() as cur:
+        cur.execute("alter session set python_connector_query_result_format='JSON'")
+        ret = cur.execute(
+            """select '"",' || '"",' || '"",' || '"",' || '"",' || 'Ofigràfic' || '"",' from TABLE(GENERATOR(ROWCOUNT => 5000)) v;"""
```
Comment (sfc-gh-mkeller) [inline test/integ/test_cursor.py:1644]: This looks like the same string factor them out to a constant to make it obvious that they were the same.

File: test/integ/test_cursor.py [hunk 9]
```diff
@@ -1618,3 +1618,30 @@ def test_multi_statement_failure(conn_cnx):
...
+        # This test case is tricky, for most of the test cases, the decoding is incorrect and can could be different
+        # on different platforms, however, due to randomness, in rare cases the decoding is indeed utf-8,
+        # the backend behavior is flaky
+        assert ret[0] in (
+            ('"","","","","",OfigrĂ\xa0fic"",',),
+            ('"","","","","",Ofigràfic"",',),
+        )
+
+    with conn_cnx(json_result_force_utf8_decoding=True) as con, con.cursor() as cur:
+        cur.execute("alter session set python_connector_query_result_format='JSON'")
```
Comment (sfc-gh-mkeller) [inline test/integ/test_cursor.py:1642]: Move this into `conn_cnx` too

**Review by sfc-gh-mkeller — APPROVED** [review]
Just small nitpicks

:ship: :ship:

File: src/snowflake/connector/connection.py (hunk 4 above)
Comment (sfc-gh-aling) [reply src/snowflake/connector/connection.py]: updated

File: test/integ/test_cursor.py (hunk 9 above)
Comment (sfc-gh-aling) [reply test/integ/test_cursor.py]: fixed

File: test/integ/test_cursor.py (hunk 8 above)
Comment (sfc-gh-aling) [reply test/integ/test_cursor.py]: fixed

File: test/integ/test_cursor.py (hunk 7 above)
Comment (sfc-gh-aling) [reply test/integ/test_cursor.py]: fixed

**Pull request merged**